src/ModelManager.cpp

```diff
@@ -444,36 +442,49 @@ void ModelManager::fetchRemoteModels() {
     if (isFetchingRemoteModels())
         return;
 
-    isFetchingRemoteModels_ = true;
-    emit fetchingRemoteModels();
-
-    QUrl url(kModelListUrl);
-    QNetworkRequest request(url);
-    QNetworkReply *reply = network_->get(request);
-    connect(reply, &QNetworkReply::finished, this, [=] {
-        switch (reply->error()) {
-            case QNetworkReply::NoError:
-                parseRemoteModels(QJsonDocument::fromJson(reply->readAll()).object());
-                break;
-            default:
-                emit error(reply->errorString());
-                break;
-        }
+    QStringList repos = repositories_.getRepos();
+    for (auto&& urlString : repos) {
+        isFetchingRemoteModels_ = true;
+        emit fetchingRemoteModels();
+
+        QUrl url(urlString);
+        QNetworkRequest request(url);
+        QNetworkReply *reply = network_->get(request);
+        connect(reply, &QNetworkReply::finished, this, [=] {
+            switch (reply->error()) {
+                case QNetworkReply::NoError:
+                    parseRemoteModels(QJsonDocument::fromJson(reply->readAll()).object());
+                    break;
+                default:
+                    QString errstr = QString("Error fetching remote repository: ") + urlString +
+                            QString("\nError code: ") + reply->errorString() +
+                            QString("\nPlease double check that the address is reachable.");
+                    emit error(errstr);
+                    break;
+            }
 
-        isFetchingRemoteModels_ = false;
-        emit fetchedRemoteModels();
+            isFetchingRemoteModels_ = false;
+            emit fetchedRemoteModels();
```

Comment:
Ugh. That. Basically the reason is that I want to `emit fetchedRemoteModels()` only after the final model is fetched, but I can't move it outside of the loop because it's called by the lambda. If I move it out of the loop it gets called before the network operation finishes in some cases.

The correct solution would be for the lambda to know when it's the last repository to fetch from the list and I will imlement that.

Reply:
Oh sorry I didn't notice that this was inside a callback! You might get away with just keeping a `int counter = repos.size();` and `if (--counter == 0) emit`. I believe all of these callbacks are executed on the same thread, so you don't even need a semaphore!

Reply:
Hmmm, but even then, I want the last completely callback... One may take longer to execute despite being first in line..

Reply:
That should still be the case, since `counter` will only be 0 when the last one calls back (assuming you bind counter by reference to the  lambda) regardless of the order they are called in. Something like this (I forgot you'd need an int on the heap because of course the outer function will go out of scope before any of the callbacks return 😅)

```cpp
QShardedPointer<int> counter(new int(repos.size()));
isFetchingRemoteModels_ = true;
for (auto &&repo : repos) {
  // do update
  connect(reply, &QNetworkRely::finished, this, [=] {
    // either success or error, doesn't really matter
    if (--(*counter) == 0) {
      isFetchingRemoteModels_ = false;
    }
  }
}
```

I'm looking at Qt's documentation wether there is something better than this, because this is a bit double! QSharedPointer already knows when all of the callbacks have finished, and keeps the same counter internally effectively! If only there was something like QScopedGuard that behaved like QSharedPointer, and would call its callback once all of its copied had gone out of scope…

Reply:
Ohhh I thought you meant passing the counter to the lambda... In this case the counter needs to be an `std::atomic<int> counter`. But yeah this will work.